File: Programmers/2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

bool check(const unordered_map<string, int>& target,
           const unordered_map<string, int>& win) 
{
    for (const auto& p : target)
    {
        if (win.count(p.first) == 0) return false;
        if (win.at(p.first) != p.second) return false;
    }
    
    return true;
}

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    
    unordered_map<string, int> target;
    for (int i = 0; i < want.size(); i++)
        target[want[i]] = number[i];

    unordered_map<string, int> win;
    
    for (int i = 0; i < 10; i++)
        win[discount[i]]++;

    if (check(target, win))
        answer++;
    
    for (int i = 10; i < discount.size(); i++)
    {
        string out = discount[i - 10];
        string in  = discount[i];

        win[out]--;
        if (win[out] == 0)
            win.erase(out);

        win[in]++;

        if (check(target, win))
            answer++;
    }    
    
    return answer;
}

File: Programmers/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        
        Map<String, Integer> target = new HashMap<>();
        for (int i = 0; i < want.length; i++)
            target.put(want[i], number[i]);

        Map<String, Integer> win = new HashMap<>();
        for (int i = 0; i < 10; i++)
            win.put(discount[i], win.getOrDefault(discount[i], 0) + 1);

        if (check(target, win)) answer++;

        for (int i = 10; i < discount.length; i++) {
            String out = discount[i - 10];
            String in  = discount[i];

            win.put(out, win.get(out) - 1);
            if (win.get(out) == 0) win.remove(out);

            win.put(in, win.getOrDefault(in, 0) + 1);

            if (check(target, win)) answer++;
        }
        
        return answer;
    }

    private boolean check(Map<String, Integer> target, Map<String, Integer> win) {
        for (String key : target.keySet()) {
            if (!win.getOrDefault(key, 0).equals(target.get(key)))
                return false;
        }
        
        return true;
    }
}